lib/versioning/regex/index.ts

```diff
@@ -0,0 +1,118 @@
+import { compare, satisfies, ltr, minSatisfying, maxSatisfying } from 'semver';
+import RE2 from 're2';
+import { VersioningApiConstructor } from '../common';
+import { GenericVersion, GenericVersioningApi } from '../loose/generic';
+
+export interface RegExpVersion extends GenericVersion {
+  /** prereleases are treated in the standard semver manner, if present */
+  prerelease: string;
+  /**
+   * compatibility, if present, are treated as a compatibility layer: we will
+   * never try to update to a version with a different compatibility.
+   */
+  compatibility: string;
+}
+
+export class RegExpVersioningApi extends GenericVersioningApi<RegExpVersion> {
+  // config is expected to be overridden by a user-specified RegExp value
+  // sample values:
+  //
+  // * emulates the "semver" configuration:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<prerelease>.*))?$')
+  // * emulates the "docker" configuration:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-(?<compatibility>.*))?$')
+  // * matches the versoining scheme used by the Python images on DockerHub:
+  //   RegExp('^(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(?<prerelease>[^.-]+)?(-(?<compatibility>.*))?$');
+  private _config: RegExp = null;
+
+  constructor(new_config: string) {
+    super();
+    if (!new_config)
+      // eslint-disable-next-line no-param-reassign
+      new_config = '^(?<major>\\d+)?$';
+
+    // without at least one of {major, minor, patch} specified in the regex,
+    // this versioner will not work properly
+    if (
+      !new_config.includes('<major>') &&
+      !new_config.includes('<minor>') &&
+      !new_config.includes('<patch>')
+    ) {
+      const error = new Error('config-validation');
+      error.configFile = new_config;
+      error.validationError =
+        'regex versionScheme needs at least one major, minor or patch group defined';
+      throw error;
+    }
+
+    // TODO: should we validate the user has not added extra unsupported
+    // capture groups?
+
+    this._config = new RE2(new_config);
```

Comment:
What happens if the regex is "unsupported" (e.g. a backtracking syntax that RE2 doesn't support)?

Reply:
or just an invalid syntax altogether

Reply:
it will throw `SyntaxError`:
![image](https://user-images.githubusercontent.com/1798109/63533955-34596a80-c50e-11e9-8089-bea52a11e1fd.png)

Should we add some tests?

Reply:
Added some tests